packages/asset-swapper/contracts/src/MakerPSMSampler.sol

```diff
@@ -0,0 +1,251 @@
+        returns (uint256[] memory makerTokenAmounts)
+    {
+        _assertValidPair(makerToken, takerToken);
+        IPSM psm = IPSM(psmInfo.psmAddress);
+        IVAT vat = IVAT(psm.vat());
+
+        uint256 numSamples = takerTokenAmounts.length;
+        makerTokenAmounts = new uint256[](numSamples);
+
+        for (uint256 i = 0; i < numSamples; i++) {
+            uint256 buyAmount = _samplePSMSell(psmInfo, makerToken, takerToken, takerTokenAmounts[i], psm, vat);
+
+            if (buyAmount == 0) {
+                break;
+            }
+            makerTokenAmounts[i] = buyAmount;
+        }
+    }
+
+    function sampleBuysFromMakerPsm(
+        MakerPsmInfo memory psmInfo,
+        address takerToken,
+        address makerToken,
+        uint256[] memory makerTokenAmounts
+    )
+        public
+        view
+        returns (uint256[] memory takerTokenAmounts)
+    {
+        _assertValidPair(makerToken, takerToken);
+        IPSM psm = IPSM(psmInfo.psmAddress);
+        IVAT vat = IVAT(psm.vat());
+
+        uint256 numSamples = makerTokenAmounts.length;
+        takerTokenAmounts = new uint256[](numSamples);
+        for (uint256 i = 0; i < numSamples; i++) {
+            uint256 sellAmount = _samplePSMBuy(psmInfo, makerToken, takerToken, makerTokenAmounts[i], psm, vat);
+
+            if (sellAmount == 0) {
+                break;
+            }
+
+            takerTokenAmounts[i] = sellAmount;
+        }
+
+    }
+
+    function _samplePSMSell(MakerPsmInfo memory psmInfo, address makerToken, address takerToken, uint256 takerTokenAmount, IPSM psm, IVAT vat)
+        private
+        view
+        returns (uint256)
+    {
+        (uint256 totalDebtInWad,,, uint256 debtCeilingInRad, uint256 debtFloorInRad) = vat.ilks(psmInfo.ilkIdentifier);
+        uint256 gemTokenBaseUnit = uint256(1e6);
+
+        if (takerToken == psmInfo.gemTokenAddress) {
+            // Simulate sellGem
+            // Selling USDC to the PSM, increasing the total debt
+            // Convert USDC 6 decimals to 18 decimals [wad]
+            uint256 takerTokenAmountInWad = takerTokenAmount.safeMul(1e12);
+
+            uint256 newTotalDebtInRad = totalDebtInWad.safeAdd(takerTokenAmountInWad).safeMul(RAY);
+
+            // PSM is too full to fit
+            if (newTotalDebtInRad >= debtCeilingInRad) {
+                return 0;
+            }
+
+            uint256 feeInWad = takerTokenAmountInWad.safeMul(psm.tin()).safeDiv(WAD);
+            uint256 makerTokenAmountInWad = takerTokenAmountInWad.safeSub(feeInWad);
+
+            return makerTokenAmountInWad;
+        } else if (makerToken == psmInfo.gemTokenAddress) {
+            // Simulate buyGem
+            // Buying USDC from the PSM, decreasing the total debt
+            // Selling DAI for USDC, already in 18 decimals [wad]
+            uint256 takerTokenAmountInWad = takerTokenAmount;
+            uint256 newTotalDebtInRad = totalDebtInWad.safeSub(takerTokenAmountInWad).safeMul(RAY);
```

Comment:
Is it possible that `totalDebtInWad < takerTokenAmountInWad`? Would we rather return 0 if that's the case?

Reply:
Yeah, I think that is a possible scenario. Good catch! 🙏  We should definitely return 0 in that case